whisper/src/message.rs

```diff
@@ -0,0 +1,476 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Whisper message parsing, handlers, and construction.
+
+use std::fmt;
+use std::time::{self, SystemTime, Duration};
+
+use bigint::hash::{H256, H512};
+use rlp::{self, DecoderError, RlpStream, UntrustedRlp};
+use smallvec::SmallVec;
+use tiny_keccak::{keccak256, Keccak};
+
+/// Work-factor proved. Takes 3 parameters: size of message, time to live,
+/// and hash.
+///
+/// Panics if size or TTL is zero.
+pub fn work_factor_proved(size: u64, ttl: u64, hash: H256) -> f64 {
+	assert!(size != 0 && ttl != 0);
+
+	let leading_zeros = {
+		let leading_zeros = hash.iter().take_while(|&&x| x == 0).count();
+		(leading_zeros * 8) + hash.get(leading_zeros + 1).map_or(0, |b| b.leading_zeros() as usize)
+	};
+	let spacetime = size as f64 * ttl as f64;
+
+	(1u64 << leading_zeros) as f64 / spacetime
+}
+
+/// A topic of a message.
+#[derive(Debug, Default, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Topic(pub [u8; 4]);
```

Comment:
The bloom algorithm is slightly different, I think. It's not too much overhead to have a separate type, and it conveys the usage better IMO